### ISSUE TITLE
Dispatch Type?

### ISSUE THREAD

**Issue by Liyw979** [joernio/joern#1895]
```
  "Flow into method defined as lambda and assigned to constant" in {
    val cpg: Cpg = code("""
        | const foo = (x, y) => {
        |   sink(x);
        | };
        |
        | foo(1, 2);
        |""".stripMargin)

    val sink = cpg.call("sink").l
    val src  = cpg.literal.code("1").l
    sink.size shouldBe 1
    src.size shouldBe 1
    sink.reachableBy(src).size shouldBe 1
  }
```
<img width="550" alt="image" src="https://user-images.githubusercontent.com/16919236/197719094-a6cff3a7-cc2c-4423-b310-148110b0ddfa.png">

I think the ```dispatchType``` of ```sink``` and ```foo``` should be ```Static```

**Comment by max-leuthaeuser** [comment]
As we do not know what `sink` actually is we set it correctly I guess (also required for call linking at a later stage). For `foo` we could set it here (`const` and directly in the same scope) - but this would  be no general solution.

**Comment by Liyw979** [comment]
I get it. Thank you!

**Comment by Liyw979** [comment]
But in Js, We must do ```this.sink``` or ```bar.sink``` to do a dynamic call. Am I right ?

**Comment by max-leuthaeuser** [comment]
Well, the semantics for `this` are rather complicated (https://www.w3schools.com/js/js_this.asp).
We did not want to rebuild all these rules here so the linking avoids looking at `this`.

See:
 - https://github.com/joernio/joern/blob/master/joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/StaticCallLinker.scala
 - https://github.com/joernio/joern/blob/master/joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/DynamicCallLinker.scala
 - https://github.com/joernio/joern/blob/master/joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/frontend/JavascriptCallLinker.scala

**Comment by Liyw979** [comment]
Ah, I mean that the ```sink(x)``` is not ```this.sink(x)``` or ```others.sink(x)``` so we can say  ```sink(x)``` is a static call.

**Comment by max-leuthaeuser** [comment]
Hm yes, I guess we could do this. I'll test that in a separate PR.

**Comment by Liyw979** [comment]
Thanks

**Comment by max-leuthaeuser** [comment]
Ok, does not work as we still have no way to determine the correct call fullnames.
Hence, we can't link statically. In your small example it could work (everything is locally in the same scope etc.) but in general the missing full names get us nowhere.

**Comment by Liyw979** [comment]
Yes, I know the both ```static linker``` and ```dynamic linker``` are depend on ```fullnames``` and they are not working because of the fullname issue.
In the issue, I just want to point out that we should set the property ```DISPATCH_TYPE``` of ```sink(x)``` to static, though this will not help the linking stage.

**Comment by max-leuthaeuser** [comment]
Sadly, that would also render the JavascriptCallLinker useless (relies on fullnames for static calls also) and method stubs are generated incorrectly as well. This (down the pipeline) causes issues with require/import resolution and more.
 
We would have to make an additional switch in all linkers/stub generators (handle/ignore the situation static call + no fullname) or run an additional pass after all of them to fix-up the calls.
@fabsx00 what do you think?

**Comment by Liyw979** [comment]
Looks complicated. I will take a deeper look at ```JavascriptCallLinker``` to see if I can help.

**Comment by max-leuthaeuser** [comment]
We could something like https://github.com/joernio/joern/pull/1937.
Lets see what @fabsx00 thinks.

**Comment by ml86** [comment]
We can not make the `sink` call static because it simply is not. Not in this example and not in general. Calls in Javascript are dispatched dynamically even if there is no instance object explicitly given. This is because calls which are just `<someIdentifier>()` like all other calls in JS are under the hood a function/method pointer.
I suggest we just leave it as it is and have passes like `JavascriptCallLinker` to associate call site to methods.

**Comment by Liyw979** [comment]
Maybe we can think of this in anther way:

- when ```DISPATCH_TYPE``` is ```static```, ```StaticCallLinker``` is used.
- when ```DISPATCH_TYPE``` is ```dynamic```, ```DynamicCallLinker``` is used.

In this case, which linker should we use to link ```sink(x)```?

**Comment by Liyw979** [comment]
Yes, JS is so dynamic. It's confusing to model dynamic behavior in JS using DISPATCH_TYPE = static.
But since the cpg of joern is a intermediate representation and we are doing translation. It's a little bit acceptable that the representation is not 100% consitence with true language behavior I guess.

**Comment by ml86** [comment]
There are other code parts which rely on the current modeling so changing that would need a good reason. Can you please reiterate why you want to treat those `sink(x)` calls as STATIC?

**Comment by Liyw979** [comment]
My starting point is that we should use ```StaticCallLinker``` to link ```sink(x)```, in case that we do figure out correct ```fullnames``` somehow **later**.

**Comment by Liyw979** [comment]
Or maybe we can add one more property ```LINKER_TYPE``` in the ```Call``` node:

- use ```DISPATCH_TYPE``` to model actual behavior
- use ```LINKER_TYPE``` as a flag to indicate which ```Caller``` will be used.

**Comment by ml86** [comment]
@Liyw979 So i take it that this is more about structuring of program logic than new features or problems in the existing code?

**Comment by Liyw979** [comment]
Yes @ml86
